### ISSUE TITLE
Use different separator then \0 for stub line

### ISSUE THREAD

**Issue by voxik** [ruby/rubygems#883]
Commit https://github.com/rubygems/rubygems/commit/652e31942e807ae4625d2cc1eb771cd296c8d1ed introduced stub lines and is using \0 as a item separator. This is unfortunate choice, since it confuses some tools. Namely, diff cannot do diff of such files, complaining about binary files, gedit opens such file just as a garbage. If .gemspec is considered text file and ruby source code, it should not contain any NULL characters.

**Comment by copiousfreetime** [comment]
This should be changeable since it only affects the `# stub:` lines in specifications. See:
* https://github.com/rubygems/rubygems/blob/master/lib/rubygems/stub_specification.rb
* https://github.com/rubygems/rubygems/blob/master/lib/rubygems/specification.rb#L2471-L2472

**Comment by bronzdoc** [comment]
What separator do you recomend @voxik? are you still interested in this feature?

**Comment by voxik** [comment]
Dunno, semicolon? Of course I still consider NULL character in text file as a bad idea. Not sure if the tools were fixed ...

**Comment by deivid-rodriguez** [comment]
I guess we could use `File::PATH_SEPARATOR` since this separator seems to be used for separating paths?

**Comment by voxik** [comment]
> I guess we could use `File::PATH_SEPARATOR` since this separator seems to be used for separating paths?

Isn't it different for *NIX/Windows?

**Comment by deivid-rodriguez** [comment]
Yes, does it matter?

**Comment by voxik** [comment]
It will make sharing the .gemspec between OSes harder?

I don't know how often this might be real issue, but there were times when I developed on Windows and deployed on Solaris. So I can imagine I would (use Bundler to) vendor the gems and I had a troubles. Dunno. That is just speculation.

**Comment by deivid-rodriguez** [comment]
Maybe? Note that these are just installed gemspec stubs, so I don't think we'd want to support share those. The only alternative I can think if is... `\0` because all the other candidates are valid characters on a filesystem path... right? So changing this would be exchanging a tiny minor annoyance with another tiny minor annoyance?

**Comment by MSP-Greg** [comment]
Re `File::PATH_SEPARATOR`, Windows Ruby and almost all recent Windows versions work fine with a forward-slash.  The only problem happens when string comparisons are done in code.  You probably already know this, but a reminder never hurts...

**Comment by deivid-rodriguez** [comment]
I think forward slash is relevant to `File::SEPARATOR`? `File::PATH_SEPARATOR` is, I believe, `:` on Unix, and `;` on Windows (because paths can have `:`).

Actually, since these seem to be relative paths, using `:` might be ok?

Also, note that this is only relevant for gemspecs using multiple `require_paths` or multiple extensions. I don't actually know of a single gem that does either of this? I assume you don't recall the specific gemspec that led you to open this issue, right @voxik?

**Comment by voxik** [comment]
> I think forward slash is relevant to `File::SEPARATOR`? `File::PATH_SEPARATOR` is, I believe, `:` on Unix, and `;` on Windows (because paths can have `:`).

Good point. I think that my original thoughts were in line with @MSP-Greg  :innocent:

> Also, note that this is only relevant for gemspecs using multiple `require_paths` or multiple extensions. I don't actually know of a single gem that does either of this? I assume you don't recall the specific gemspec that led you to open this issue, right @voxik?

[json](https://rubygems.org/gems/json) has two extensions, so that might be the case.

**Comment by MSP-Greg** [comment]
I like to think I'm capable of reading at a minimum of a semi-literate level, but I haven't demonstrated that here.  Sorry.

**Comment by deivid-rodriguez** [comment]
No, problem.

I was going to close this ticket as won't fix, but then I noticed that these paths are relative, so using ":" sounds like a perfect fit to me.

**Comment by eregon** [comment]
Seems I forgot to comment here, but :+1: from me (from https://github.com/rubygems/rubygems/pull/4199#issuecomment-1207212755).
Those files are meant to be human-readable, and so they should not contain `\0` which is I believe is universally known as something only for binary/machine-only-readable files.

**Comment by hsbt** [comment]
Closing as won't fix. Installed gemspecs are generated files that RubyGems reads, not files to read by hand. Content addressable gems, which land in an upcoming release, will put a content address in the third stub column instead of a platform. `diff -a` or `git diff --text` handles the NUL byte, and `gem specification NAME --ruby` compares two specifications properly, accepting both an installed gem name and a `.gem` file.
